usrsctplib/netinet/sctp_pcb.c

```diff
@@ -5237,6 +5225,21 @@ sctp_remove_net(struct sctp_tcb *stcb, struct sctp_nets *net)
 		sctp_free_remote_addr(stcb->asoc.alternate);
 		stcb->asoc.alternate = NULL;
 	}
+	sctp_timer_stop(SCTP_TIMER_TYPE_SEND, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_9);
+	sctp_timer_stop(SCTP_TIMER_TYPE_INIT, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_10);
+	sctp_timer_stop(SCTP_TIMER_TYPE_SHUTDOWN, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_11);
+	sctp_timer_stop(SCTP_TIMER_TYPE_COOKIE, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_12);
+	sctp_timer_stop(SCTP_TIMER_TYPE_SHUTDOWNACK, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_13);
+	sctp_timer_stop(SCTP_TIMER_TYPE_PATHMTURAISE, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_14);
+	sctp_timer_stop(SCTP_TIMER_TYPE_HEARTBEAT, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_15);
+	net->removed = 1; /* Prevents timers from being restarted. */
```

Comment:
I think stopping the HB and PMTU timer here is actually the correct place. However, we should not stop the other timers here, since, if for example the SEND timer is not running anymore, no retransmissions for the DATA chunks sent on the now removed net will happen. I committed a corresponding fix.

Reply:
Can you summarize how retransmissions work? I'm now worried that the `net->removed` logic is also flawed; should it be possible to restart the SEND timer after removing?

I'm doing this because the old behavior was "stop timers when there are zero references". This doesn't work since the timers themselves now add references.

So I assumed (pretty naively I now realize) that I could just stop the timers here. If that's not possible, then we still have the same issue: a timer holding a reference could keep the net alive when it otherwise would have been freed...

If we want exactly the same behavior as before, maybe we could use two reference counts as I mentioned previously.

When the main reference count is reduced to 0, we'd stop the timers (as they were stopped before), but would only free the memory if the timer reference count is also 0.

The new problem would be that we'd be setting the `removed`/"about to be freed" flag inside `sctp_free_remote_addr`, which would be racy since I'm assuming there is no guarantee that the association lock is held wherever that's called.

So I'm somewhat at a loss now. We could address that problem by adding a `NETKILL` timer but now things are getting pretty messy.

I went ahead and rebased, but I want to be sure this is watertight before merging. Do you have any ideas?

Reply:
> Can you summarize how retransmissions work? I'm now worried that the `net->removed` logic is also flawed; should it be possible to restart the SEND timer after removing?

If you send a DATA chunk on a `net`, you store a pointer to it in `whoTo` and start a timer for it. Assuming we have multiple `net`s, when the timer runs off, we choose a different net, retransmit the DATA chunk there and start a new timer on the other `net`.

If we remove a `net`, and there is outstanding data, we need to wait until the timer runs off. In the old code, we would not free the net since there was a ref count on all outstanding DATA chunks. But I think this should work with the new stuff the same way.

I guess this can be optimized be moving chunks over to a different `net`, but that can be a separate commit.
This also does not apply to the WebRTC case, since dynamic address configuration is not used there.
> 
> I'm doing this because the old behavior was "stop timers when there are zero references". This doesn't work since the timers themselves now add references.
We would remove the net, after the timer is running off. That is the same as in the old case, since we would have ref counts on the DATA chunks.
> 
> So I assumed (pretty naively I now realize) that I could just stop the timers here. If that's not possible, then we still have the same issue: a timer holding a reference could keep the net alive when it otherwise would have been freed...

I don't see that as an issue and the behaviour should be the same.

> 
> If we want exactly the same behavior as before, maybe we could use two reference counts as I mentioned previously.
> 
> When the main reference count is reduced to 0, we'd stop the timers (as they were stopped before), but would only free the memory if the timer reference count is also 0.

When we are removing the `net`, it does not make any sense to send packets via the `net`. So we should stop the HB and PMTU timer right away. That is why I committed that. The DATA retransmission must be kept running... Or me move the data away...

> 
> The new problem would be that we'd be setting the `removed`/"about to be freed" flag inside `sctp_free_remote_addr`, which would be racy since I'm assuming there is no guarantee that the association lock is held wherever that's called.

I think the STCB lock should be hold when modifying any `net`. Do you have a place in mind, where it is not held?

> 
> So I'm somewhat at a loss now. We could address that problem by adding a `NETKILL` timer which would repeatedly attempt to stop the timers, but now things are getting pretty messy.
> 
> I went ahead and rebased, but I want to be sure this is watertight before merging. Do you have any ideas?

Thanks for re-basing. I'm trying to commit things in small pieces to reduce the diff and get a better view regarding the the changes. I think ref counting should work. I'm also putting the changes to the FreeBSD kernel, since we have there some features for detecting memory loss or locking issues...

Reply:
>  Do you have a place in mind, where it is not held?
No, I just wasn't sure and didn't look exhaustively. I assumed that since the atomic increment was being used, that was the only thing guaranteeing thread safety.

Thanks for your explanation and for bearing with me.